File: scripts/wait-for-benchmark-slot.py
#!/usr/bin/env python3

import argparse
import email.utils
import json
import os
import sys
import time
import urllib.error
import urllib.request
from datetime import datetime, timezone
from typing import Iterable


BEHAVIOR_WORKFLOW_NAMES = {
    "Behavior Benchmark Smoke",
    "Behavior Benchmark Full",
    "Behavior Benchmark Subagents Smoke",
}

ACTIVE_STATUSES = {"queued", "in_progress", "waiting", "pending", "requested"}


RATE_LIMIT_BODY_MARKERS = frozenset(["rate_limit_exceeded", "rate limit exceeded", "rate limit"])

def _read_body_once(exc: urllib.error.HTTPError) -> str:
    """Cache the response body on the exception so we can read it more than once."""
    cached = getattr(exc, "_body_cache", None)
    if cached is not None:
        return cached
    body = exc.read().decode("utf-8", errors="replace")
    exc._body_cache = body  # type: ignore[attr-defined]
    return body


def is_github_rate_limit(exc: urllib.error.HTTPError) -> bool:
    """Return True if this HTTP 403 is a GitHub API rate-limit response."""
    if exc.code != 403:
        return False
    remaining = exc.headers.get("X-RateLimit-Remaining", "")
    if remaining == "0":
        return True
    body = _read_body_once(exc)
    return any(marker in body.lower() for marker in RATE_LIMIT_BODY_MARKERS)


def parse_retry_after(value: str | None) -> int | None:
    """Parse a Retry-After header (seconds or HTTP-date). Return None on failure."""
    if not value:
        return None
    candidate = value.strip()
    if not candidate:
        return None
    try:
        seconds = int(candidate)
    except ValueError:
        seconds = None
    if seconds is not None:
        return max(0, seconds)
    try:
        when = email.utils.parsedate_to_datetime(candidate)
    except (TypeError, ValueError):
        return None
    if when is None:
        return None
    if when.tzinfo is None:
        when = when.replace(tzinfo=timezone.utc)
    delta = (when - datetime.now(timezone.utc)).total_seconds()
    return max(0, int(delta))


def handle_rate_limit(exc: urllib.error.HTTPError) -> int | None:
    """Sleep for the Retry-After period on a rate-limit hit; return None to retry."""
    parsed = parse_retry_after(exc.headers.get("Retry-After"))
    wait_seconds = parsed if parsed is not None else 60
    print(
        f"GitHub API rate limit hit (HTTP 403). "
        f"Retrying after {wait_seconds}s (Retry-After header, default 60).",
        file=sys.stderr,
    )
    time.sleep(wait_seconds)
    return None  # signal caller to continue/retry


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--current-run-id", type=int, required=True)
    parser.add_argument("--head-sha", required=True)
    parser.add_argument("--max-active", type=int, default=2)
    parser.add_argument("--poll-seconds", type=int, default=15)
    parser.add_argument("--timeout-seconds", type=int, default=3600)
    parser.add_argument("--repo", default=os.environ.get("GITHUB_REPOSITORY", ""))
    parser.add_argument("--api-url", default=os.environ.get("GITHUB_API_URL", "https://api.github.com"))
    return parser.parse_args()


def build_request(url: str, token: str) -> urllib.request.Request:
    return urllib.request.Request(
        url,
        headers={
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {token}",
            "X-GitHub-Api-Version": "2022-11-28",
            "User-Agent": "claude-crew-benchmark-slot-gate",
        },
    )


def fetch_active_behavior_runs(*, api_url: str, repo: str, token: str, head_sha: str) -> list[dict]:
    url = f"{api_url}/repos/{repo}/actions/runs?per_page=100"
    request = build_request(url, token)
    with urllib.request.urlopen(request, timeout=30) as response:
        payload = json.load(response)

    runs = payload.get("workflow_runs", [])
    selected: list[dict] = []
    for run in runs:
        if run.get("name") not in BEHAVIOR_WORKFLOW_NAMES:
            continue
        if run.get("head_sha") != head_sha:
            continue
        if run.get("status") not in ACTIVE_STATUSES:
            continue
        selected.append(run)
    return selected


def order_active_runs(runs: Iterable[dict]) -> list[dict]:
    return sorted(
        runs,
        key=lambda run: (
            run.get("created_at", ""),
            int(run.get("id", 0)),
        ),
    )


def current_run_has_slot(*, current_run_id: int, runs: Iterable[dict], max_active: int) -> tuple[bool, list[int]]:
    ordered = order_active_runs(runs)
    allowed_run_ids = [int(run["id"]) for run in ordered[:max_active]]
    return current_run_id in allowed_run_ids, allowed_run_ids


def handle_transient_error(exc: Exception, attempt: int) -> bool:
    """Log a transient error and optionally retry. Returns True if we should retry."""
    max_retries = 5
    if attempt >= max_retries:
        return False
    delay = min(2 ** attempt, 60)
    print(f"Transient error: {exc}. Retry {attempt + 1}/{max_retries} after {delay}s.", file=sys.stderr)
    time.sleep(delay)
    return True


def main() -> int:
    args = parse_args()
    token = os.environ.get("GITHUB_TOKEN", "")
    if not args.repo:
        print("GITHUB_REPOSITORY or --repo is required", file=sys.stderr)
        return 2
    if not token:
        print("GITHUB_TOKEN is required", file=sys.stderr)
        return 2

    deadline = time.monotonic() + args.timeout_seconds
    transient_attempt = 0
    while True:
        try:
            active_runs = fetch_active_behavior_runs(
                api_url=args.api_url,
                repo=args.repo,
                token=token,
                head_sha=args.head_sha,
            )
            transient_attempt = 0
        except urllib.error.HTTPError as exc:
            if is_github_rate_limit(exc):
                result = handle_rate_limit(exc)
                if result is not None:
                    return result
                continue  # retry after sleeping
            if exc.code in {500, 502, 503, 504}:
                if handle_transient_error(exc, transient_attempt):
                    transient_attempt += 1
                    continue
            try:
                body_snippet = _read_body_once(exc)[:500]
            except Exception:  # noqa: BLE001 - body may already be consumed
                body_snippet = ""
            if body_snippet:
                print(
                    f"GitHub API request failed with HTTP {exc.code}: {body_snippet}",
                    file=sys.stderr,
                )
            else:
                print(f"GitHub API request failed with HTTP {exc.code}", file=sys.stderr)
            return 1
        except urllib.error.URLError as exc:
            if handle_transient_error(exc, transient_attempt):
                transient_attempt += 1
                continue
            print(f"GitHub API request failed: {exc}", file=sys.stderr)
            return 1

        has_slot, allowed_ids = current_run_has_slot(
            current_run_id=args.current_run_id,
            runs=active_runs,
            max_active=args.max_active,
        )
        active_ids = [int(run["id"]) for run in order_active_runs(active_runs)]
        print(
            f"Active benchmark workflow runs for {args.head_sha[:12]}: {active_ids}. "
            f"Allowed now: {allowed_ids}."
        )
        if has_slot:
            print(f"Run {args.current_run_id} has a benchmark slot.")
            return 0
        if time.monotonic() >= deadline:
            print(
                f"Timed out waiting for benchmark slot after {args.timeout_seconds}s. "
                f"Current run {args.current_run_id} never entered the first {args.max_active}.",
                file=sys.stderr,
            )
            return 1
        time.sleep(args.poll_seconds)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/wait-for-benchmark-slot.mjs
#!/usr/bin/env node
// wait-for-benchmark-slot: polls the GitHub Actions API and waits until the
// current run is among the `max-active` oldest active benchmark runs.
// Node port of scripts/wait-for-benchmark-slot.py — no Python. Uses global
// fetch (Node 22+); testable via the exported `clock` and `deps` objects.
import { isMain, githubHeaders, parseRetryAfter } from './bench/lib.mjs';

export { parseRetryAfter };

const BEHAVIOR_WORKFLOW_NAMES = new Set([
  'Behavior Benchmark Smoke',
  'Behavior Benchmark Full',
  'Behavior Benchmark Subagents Smoke',
]);
const ACTIVE_STATUSES = new Set(['queued', 'in_progress', 'waiting', 'pending', 'requested']);
const RATE_LIMIT_BODY_MARKERS = ['rate_limit_exceeded', 'rate limit exceeded', 'rate limit'];

export class GithubHttpError extends Error {
  constructor(code, headers, body = '') {
    super(`HTTP ${code}`);
    this.code = code;
    this.headers = headers;
    this.body = body;
  }
}

export function orderActiveRuns(runs) {
  return [...runs].sort((a, b) => {
    const ca = a.created_at || '';
    const cb = b.created_at || '';
    if (ca !== cb) return ca < cb ? -1 : 1;
    return Number(a.id || 0) - Number(b.id || 0);
  });
}

export function currentRunHasSlot({ currentRunId, runs, maxActive }) {
  const ordered = orderActiveRuns(runs);
  const allowed = ordered.slice(0, maxActive).map((r) => Number(r.id));
  return [allowed.includes(Number(currentRunId)), allowed];
}

export function isGithubRateLimit(err) {
  if (err.code !== 403) return false;
  const remaining = err.headers && err.headers.get ? err.headers.get('X-RateLimit-Remaining') : '';
  if (remaining === '0') return true;
  const body = (err.body || '').toLowerCase();
  return RATE_LIMIT_BODY_MARKERS.some((m) => body.includes(m));
}

export function buildRequest(url, token) {
  return { url, headers: githubHeaders(token) };
}

export async function fetchActiveBehaviorRuns({ apiUrl, repo, token, headSha }) {
  const url = `${apiUrl}/repos/${repo}/actions/runs?per_page=100`;
  const r = await fetch(url, { headers: githubHeaders(token) });
  if (!r.ok) {
    let body = '';
    try { body = await r.text(); } catch {}
    throw new GithubHttpError(r.status, r.headers, body);
  }
  const payload = await r.json();
  const runs = payload.workflow_runs || [];
  return runs.filter((run) =>
    BEHAVIOR_WORKFLOW_NAMES.has(run.name) &&
    run.head_sha === headSha &&
    ACTIVE_STATUSES.has(run.status));
}

const realClock = {
  sleep: (seconds) => {
    const buf = new Int32Array(new SharedArrayBuffer(4));
    Atomics.wait(buf, 0, 0, Math.max(0, seconds) * 1000);
  },
  monotonic: () => performance.now() / 1000,
};

export const clock = { sleep: realClock.sleep, monotonic: realClock.monotonic };

export function handleRateLimit(err) {
  const parsed = parseRetryAfter(err.headers && err.headers.get ? err.headers.get('Retry-After') : null);
  const waitSeconds = parsed !== null ? parsed : 60;
  process.stderr.write(`GitHub API rate limit hit (HTTP 403). Retrying after ${waitSeconds}s (Retry-After header, default 60).\n`);
  clock.sleep(waitSeconds);
  return null;
}

export function handleTransientError(err, attempt) {
  const maxRetries = 5;
  if (attempt >= maxRetries) return false;
  const delay = Math.min(2 ** attempt, 60);
  process.stderr.write(`Transient error: ${err && err.message ? err.message : err}. Retry ${attempt + 1}/${maxRetries} after ${delay}s.\n`);
  clock.sleep(delay);
  return true;
}

export const deps = { fetchActiveBehaviorRuns, handleRateLimit };

function parseArgs(argv) {
  const out = {
    currentRunId: null, headSha: null, maxActive: 2, pollSeconds: 15, timeoutSeconds: 3600,
    repo: process.env.GITHUB_REPOSITORY || '', apiUrl: process.env.GITHUB_API_URL || 'https://api.github.com',
  };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--current-run-id') out.currentRunId = Number.parseInt(argv[++i], 10);
    else if (a === '--head-sha') out.headSha = argv[++i];
    else if (a === '--max-active') out.maxActive = Number.parseInt(argv[++i], 10);
    else if (a === '--poll-seconds') out.pollSeconds = Number.parseInt(argv[++i], 10);
    else if (a === '--timeout-seconds') out.timeoutSeconds = Number.parseInt(argv[++i], 10);
    else if (a === '--repo') out.repo = argv[++i];
    else if (a === '--api-url') out.apiUrl = argv[++i];
    else { process.stderr.write(`unknown argument: ${a}\n`); process.exit(2); }
  }
  if (out.currentRunId === null || !out.headSha) { process.stderr.write('Usage: wait-for-benchmark-slot.mjs --current-run-id N --head-sha SHA [...]\n'); process.exit(2); }
  return out;
}

export async function main() {
  const args = parseArgs(process.argv.slice(2));
  const token = process.env.GITHUB_TOKEN || '';
  if (!args.repo) { process.stderr.write('GITHUB_REPOSITORY or --repo is required\n'); return 2; }
  if (!token) { process.stderr.write('GITHUB_TOKEN is required\n'); return 2; }

  const deadline = clock.monotonic() + args.timeoutSeconds;
  let transientAttempt = 0;
  while (true) {
    let activeRuns;
    try {
      activeRuns = await deps.fetchActiveBehaviorRuns({
        apiUrl: args.apiUrl, repo: args.repo, token, headSha: args.headSha,
      });
      transientAttempt = 0;
    } catch (err) {
      if (err instanceof GithubHttpError) {
        if (isGithubRateLimit(err)) {
          const result = deps.handleRateLimit(err);
          if (result !== null) return result;
          continue;
        }
        if ([500, 502, 503, 504].includes(err.code)) {
          if (handleTransientError(err, transientAttempt)) { transientAttempt++; continue; }
        }
        const snippet = (err.body || '').slice(0, 500);
        if (snippet) process.stderr.write(`GitHub API request failed with HTTP ${err.code}: ${snippet}\n`);
        else process.stderr.write(`GitHub API request failed with HTTP ${err.code}\n`);
        return 1;
      }
      if (handleTransientError(err, transientAttempt)) { transientAttempt++; continue; }
      process.stderr.write(`GitHub API request failed: ${err && err.message ? err.message : err}\n`);
      return 1;
    }

    const [hasSlot, allowedIds] = currentRunHasSlot({
      currentRunId: args.currentRunId, runs: activeRuns, maxActive: args.maxActive,
    });
    const activeIds = orderActiveRuns(activeRuns).map((r) => Number(r.id));
    process.stdout.write(`Active benchmark workflow runs for ${args.headSha.slice(0, 12)}: ${JSON.stringify(activeIds)}. Allowed now: ${JSON.stringify(allowedIds)}.\n`);
    if (hasSlot) {
      process.stdout.write(`Run ${args.currentRunId} has a benchmark slot.\n`);
      return 0;
    }
    if (clock.monotonic() >= deadline) {
      process.stderr.write(`Timed out waiting for benchmark slot after ${args.timeoutSeconds}s. Current run ${args.currentRunId} never entered the first ${args.maxActive}.\n`);
      return 1;
    }
    clock.sleep(args.pollSeconds);
  }
}

if (isMain(import.meta.url)) {
  main().then((code) => process.exit(code ?? 0));
}